sql/core/src/main/scala/org/apache/spark/sql/execution/ReplaceHashWithSortAgg.scala

```diff
@@ -0,0 +1,139 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import org.apache.spark.sql.catalyst.expressions.SortOrder
+import org.apache.spark.sql.catalyst.expressions.aggregate.{Complete, Final, Partial}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.aggregate.HashAggregateExec
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * Replace [[HashAggregateExec]] with [[SortAggregateExec]] in the spark plan if:
+ *
+ * 1. The plan is a pair of partial and final [[HashAggregateExec]], and the child of partial
+ *    aggregate satisfies the sort order of corresponding [[SortAggregateExec]].
+ * or
+ * 2. The plan is a [[HashAggregateExec]], and the child satisfies the sort order of
+ *    corresponding [[SortAggregateExec]].
+ *
+ * Examples:
+ * 1. aggregate after join:
+ *
+ *  HashAggregate(t1.i, SUM, final)
+ *               |                         SortAggregate(t1.i, SUM, complete)
+ * HashAggregate(t1.i, SUM, partial)   =>                |
+ *               |                            SortMergeJoin(t1.i = t2.j)
+ *    SortMergeJoin(t1.i = t2.j)
+ *
+ * 2. aggregate after sort:
+ *
+ * HashAggregate(t1.i, SUM, partial)         SortAggregate(t1.i, SUM, partial)
+ *               |                     =>                  |
+ *           Sort(t1.i)                                Sort(t1.i)
+ *
+ * [[HashAggregateExec]] can be replaced when its child satisfies the sort order of
+ * corresponding [[SortAggregateExec]]. [[SortAggregateExec]] is faster in the sense that
+ * it does not have hashing overhead of [[HashAggregateExec]].
+ */
+object ReplaceHashWithSortAgg extends Rule[SparkPlan] {
+  def apply(plan: SparkPlan): SparkPlan = {
+    if (!conf.getConf(SQLConf.REPLACE_HASH_WITH_SORT_AGG_ENABLED)) {
+      plan
+    } else {
+      replaceHashAgg(plan)
+    }
+  }
+
+  /**
+   * Replace [[HashAggregateExec]] with [[SortAggregateExec]].
+   */
+  private def replaceHashAgg(plan: SparkPlan): SparkPlan = {
+    plan.transformDown {
+      case hashAgg: HashAggregateExec =>
+        val sortAgg = hashAgg.toSortAggregate
+        hashAgg.child match {
+          case partialAgg: HashAggregateExec if isPartialAgg(partialAgg, hashAgg) =>
+            if (SortOrder.orderingSatisfies(
+                partialAgg.child.outputOrdering, sortAgg.requiredChildOrdering.head)) {
+              sortAgg.copy(
+                aggregateExpressions = sortAgg.aggregateExpressions.map(_.copy(mode = Complete)),
```

Comment:
is it always right? I think we also need to check the output partitioning to see if we can eliminate the partial agg.

An example is `df.sortWithinPartitions`. It does not cluster the data, just sort it within each partition.

Reply:
@cloud-fan - I don't think we need to check output partitioning, as we are matching a pair of final and partial hash agg, without shuffle in between:

```
  HashAggregate(final)
          |                                       SortAggregate(complete)
HashAggregate(partial)             =>                    |
          |                                            child
        child 
```

So `child` must already have proper output partitioning for `SortAggregate`, o.w. it cannot satisfy original `HashAggregate(final)`'s required distribution.

Reply:
ah ok, if there is a shuffle in the middle, we can't optimize? This looks quite limited, as having a shuffle in the middle is very common.

Reply:
> if there is a shuffle in the middle, we can't optimize?

We can, and the rule here also does pattern matching for single `HashAggregate` below. I added a unit test case in `ReplaceHashWithSortAggSuite.scala` to demonstrate replacing partial aggregate - `"replace partial hash aggregate with sort aggregate"`. But I think it would be rare to be able to replace final aggregate (though this rule also covers it), as final aggregate is almostly always immediately after a shuffle, so there's no sort ordering before final aggregate.

Spark native shuffle does not guarantee any sort orders, for Cosco (a remote shuffle service we are running in-house), we support sorted shuffle, so final aggregate can also be possible to replace.